src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs

```diff
@@ -81,20 +93,38 @@ public Parser(in GeneratorExecutionContext executionContext)
                 _executionContext = executionContext;
                 _metadataLoadContext = new MetadataLoadContextInternal(executionContext.Compilation);
 
-                _ienumerableType = _metadataLoadContext.Resolve(typeof(IEnumerable));
-                _listOfTType = _metadataLoadContext.Resolve(typeof(List<>));
-                _dictionaryType = _metadataLoadContext.Resolve(typeof(Dictionary<,>));
-
-                _booleanType = _metadataLoadContext.Resolve(typeof(bool));
-                _byteArrayType = _metadataLoadContext.Resolve(typeof(byte[]));
-                _charType = _metadataLoadContext.Resolve(typeof(char));
-                _dateTimeType = _metadataLoadContext.Resolve(typeof(DateTime));
-                _dateTimeOffsetType = _metadataLoadContext.Resolve(typeof(DateTimeOffset));
-                _guidType = _metadataLoadContext.Resolve(typeof(Guid));
-                _nullableOfTType = _metadataLoadContext.Resolve(typeof(Nullable<>));
-                _stringType = _metadataLoadContext.Resolve(typeof(string));
-                _uriType = _metadataLoadContext.Resolve(typeof(Uri));
-                _versionType = _metadataLoadContext.Resolve(typeof(Version));
+                _ilistOfTType = ResolveType(SpecialType.System_Collections_Generic_IList_T);
+                _icollectionOfTType = ResolveType(SpecialType.System_Collections_Generic_ICollection_T);
+                _ienumerableOfTType = ResolveType(SpecialType.System_Collections_Generic_IEnumerable_T);
+                _ienumerableType = ResolveType(SpecialType.System_Collections_IEnumerable);
+
+                _listOfTType = ResolveType(typeof(List<>).FullName!);
+                _dictionaryType = ResolveType(typeof(Dictionary<,>).FullName!);
+                _idictionaryOfTKeyTValueType = ResolveType(typeof(IDictionary<,>).FullName!);
+                _ireadonlyDictionaryType = ResolveType(typeof(IReadOnlyDictionary<,>).FullName!);
+                _isetType = ResolveType(typeof(ISet<>).FullName!);
+                _stackOfTType = ResolveType(typeof(Stack<>).FullName!);
+                _queueOfTType = ResolveType(typeof(Queue<>).FullName!);
+                _concurrentStackType = ResolveType(typeof(ConcurrentStack<>).FullName!);
+                _concurrentQueueType = ResolveType(typeof(ConcurrentQueue<>).FullName!);
+                _idictionaryType = ResolveType(typeof(IDictionary).FullName!);
+                _ilistType = ResolveType(typeof(IList).FullName!);
+                _stackType = ResolveType(typeof(Stack).FullName!);
```

Comment:
Am I right to assume that that we're resolving these type instances at runtime to facilitate potential trimming scenaria? If that's the case, doesn't the fact that we're introducing a static dependency to that type via the `typeof()` declarations in the RHS invalidate that benefit?

Reply:
This (and every thing else in this assembly) is all compile-time inspection of types to figure out what code to generate, so the references made here don't cause things to be rooted in the trimmed app.

Reply:
Got it. Why not just write something like `_stackType = typeof(Stack)` directly?

Reply:
The Roslyn-compilation type system is complex. Basically the types in the generator-implementation dll are not the same as the types in the input compilation, so we have to resolve them each time the generator runs. cc @chsienki who might have a better answer.